### PR TITLE
Add DevServices

### PR DESCRIPTION
**Pull request by mcruzdev**
This pull request aims to fix the following issues:

#147

### CODE REVIEW

**Comment by zhfeng** [comment]
Thanks @mcruzdev for this contribution. Please ping me if it is reay for reviewing.

**Comment by mcruzdev** [comment]
@zhfeng Surely, I am testing locally here!

File: deployment/pom.xml [hunk 1]
```diff
@@ -40,7 +40,11 @@
             <artifactId>protobuf-java-util</artifactId>
             <scope>test</scope>
         </dependency>
-
+
```
Comment (mcruzdev) [inline deployment/pom.xml:43]:
```suggestion
```

**Comment by mcruzdev** [comment]
Hi @zhfeng I think that you can review now! Thank you!

File: deployment/src/main/java/io/quarkiverse/dapr/deployment/DevServicesDaprProcessor.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+                for (Component component : components) {
+                    dapr = dapr.withComponent(component);
+                }
+            } catch (IOException e) {
+                LOGGER.warn("Was not possible to add custom components to Dapr Sidecar", e);
+            }
+
+            final Supplier<Integer> getGrpcPort = () -> !config.shared ? config.grpcPort : DAPRD_GRPC_PORT;
+            final Supplier<Integer> getHttpPort = () -> !config.shared ? config.httpPort : DAPRD_HTTP_PORT;
+
```
Comment (mcruzdev) [inline deployment/src/main/java/io/quarkiverse/dapr/deployment/DevServicesDaprProcessor.java:164]:
```suggestion
// This is necessary when the user wants to use Dev Services shared
```

File: deployment/src/main/java/io/quarkiverse/dapr/deployment/DevServicesDaprProcessor.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                for (Component component : components) {
+                    dapr = dapr.withComponent(component);
+                }
+            } catch (IOException e) {
+                LOGGER.warn("Was not possible to add custom components to Dapr Sidecar", e);
+            }
+
+            final Supplier<Integer> getGrpcPort = () -> !config.shared ? config.grpcPort : DAPRD_GRPC_PORT;
+            final Supplier<Integer> getHttpPort = () -> !config.shared ? config.httpPort : DAPRD_HTTP_PORT;
+// This is necessary when the user wants to use Dev Services shared
```
Comment (mcruzdev) [inline deployment/src/main/java/io/quarkiverse/dapr/deployment/DevServicesDaprProcessor.java:164]:
```suggestion
                // This is necessary when the user wants to use Dev Services shared
```

**Comment by mcruzdev** [comment]
Hi @zhfeng, I think that you can review now.

File: docs/modules/ROOT/pages/includes/devservices.adoc [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+== Shared Dapr
+
+In some scenarios, when you are running the application locally, you need to share the Dapr Sidecar between applications. Dev Services for Dapr implements a _service discovery_ mechanism for your multiple Quarkus applications running in _dev_ mode to share a single Dapr Sidecar.
+
+[NOTE]
+Dev Services for Dapr starts the container with the `quarkus-dev-service-dapr` label, which is used to identify the container.
+
+If you need multiple (shared) Dapr Sidecars, you can configure the `quarkus.dapr.devservices.service-name` attribute and indicate the sidecar name. It looks for a container with the same value or starts a new one if none can be found. The default service name is `dapr`.
+
+Sharing is enabled by default in dev mode but disabled in test mode. You can disable sharing with `quarkus.dapr.devservices.shared=false`.
```
Comment (mcruzdev) [inline docs/modules/ROOT/pages/includes/devservices.adoc:18]:
```suggestion
Sharing is enabled by default in dev mode. You can disable sharing with `quarkus.dapr.devservices.shared=false`.
```

File: docs/modules/ROOT/pages/includes/devservices.adoc [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+[NOTE]
+Dev Services for Dapr starts the container with the `quarkus-dev-service-dapr` label, which is used to identify the container.
+
+If you need multiple (shared) Dapr Sidecars, you can configure the `quarkus.dapr.devservices.service-name` attribute and indicate the sidecar name.
+It looks for a container with the same value or starts a new one if none can be found.
+The default service name is `dapr`.
+
+<<<<<<< HEAD
+Sharing is enabled by default in dev mode. You can disable sharing with `quarkus.dapr.devservices.shared=false`.
```
Comment (mcruzdev) [inline docs/modules/ROOT/pages/includes/devservices.adoc:22]: @salaboy

**Comment by zhfeng** [comment]
@mcruzdev very nice work! and I just add some comments.

Can we only fix one issue per PR? It could be much clearer for the mainainers to understand.

Thanks a lot!

**Comment by salaboy** [comment]
@zhfeng thanks for the feedback!

File: runtime/src/main/java/io/quarkiverse/dapr/serializer/JacksonDaprObjectSerializer.java [hunk 6]
```diff
@@ -48,6 +48,9 @@ public byte[] serialize(Object state) throws IOException {
 
     @Override
     public <T> T deserialize(byte[] data, TypeRef<T> type) throws IOException {
+        if (data.length == 0) {
+            return null;
+        }
```
[the remark this answers is not in the document]
Comment (mcruzdev) [reply runtime/src/main/java/io/quarkiverse/dapr/serializer/JacksonDaprObjectSerializer.java:53]: Makes sense! Thanks!

File: deployment/src/main/java/io/quarkiverse/dapr/deployment/DevServicesDaprProcessor.java [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+            LOGGER.debug("Not starting Dev Services for Dapr, as it has been disabled in the config.");
+            return null;
+        }
+
+        if (!dockerStatusBuildItem.isDockerAvailable()) {
+            LOGGER.warn("Docker isn't working.");
+            return null;
+        }
+
+        DaprContainer dapr = new DaprContainer("daprio/daprd:1.12.2")
```
[the remark this answers is not in the document]
Comment (mcruzdev) [reply deployment/src/main/java/io/quarkiverse/dapr/deployment/DevServicesDaprProcessor.java]: Done @zhfeng thank you, this behavior is better!

**Review by zhfeng — APPROVED** [review]

**Comment by zhfeng** [comment]
I will start a new release soon. Thanks all of you!

**Comment by zhfeng** [comment]
@all-contributors please add @mcruzdev as code!

**Pull request merged**